### ISSUE TITLE
Support recursive fixpoint serialzation.

### ISSUE THREAD

**Issue by DhashS** [sksamuel/avro4s#307]
In the example on the README, a simple algebra of kind `*` is used to demonstrate recursive schema formation, leading to the possibility to serialize recursive definitions written like so:

```scala
sealed trait Expr
case class Mul(lhs: Expr, rhs: Expr) extends Expr
case class Lit(value: Int) extends Expr
```

However, it is possible to define algebras of kind `* -> *`, often used due to their greater expressive power, representing structures with `Fix` `Mu`, and `Nu`.

```scala
sealed trait Expr[T]
case class Mul[T](lhs: T, rhs:T) extends Expr[T]
case class Lit[T](value: Int) extends Expr[T]
```

To actually make this work, i had to write some weird code.

```scala
/*
  This implicit conversion enables Fixpoint type serialzation.
  Avro does not allow mutually recursive type definitions (https://issues.apache.org/jira/browse/AVRO-530)
  Avro does not allow named unions (https://issues.apache.org/jira/browse/AVRO-248).
  Using the approach in AVRO-530 as a temporary workaround, we reorder the schema to have no forward dependencies
   */
  implicit def SchemaForFix[F[_]: Functor](implicit ev: SchemaFor[F[REPLACE.type]]): SchemaFor[Fix[F]] = {
    val replaceNamespace = REPLACE.getClass.getName.split('$').head
    val replaceName      = REPLACE.getClass.getSimpleName.replaceAllLiterally("$", "")
    val recordsInF              = read(SchemaFor[F[REPLACE.type]].schema.toString).arr
    val names            = recordsInF.map(_("name").str).toList.map(_.toString)
    //One of the forms REPLACE can take, since it can get serialized as a case class by avro4s for some reason
    val replaceStr =
      s"""{"type":"record","name":"$replaceName","namespace":"$replaceNamespace","fields":[]}""".stripMargin
    //The straightforward type replace
    val replaceStr2 =
      s""""$replaceNamespace.$replaceName"""".stripMargin
    def replacePlaceholderTermsInFieldsWithUnion(inputObject: ujson.Js, accum: ujson.Js): ujson.Js = {
      // To figure out where to splice into the list
      val lastName     = names.indexOf(accum("name").str)
      val definedSoFar = "[\"" + names.take(lastName).mkString("\",\"") + "\"," + accum.toString + "]"
      val replacementArr = inputObject("fields").arr.toList.foldLeft(List.empty[ujson.Js.Value])({
        case (orderedFieldTypes, field) => {
          val replacement =
            field.toString.replaceAllLiterally(replaceStr, definedSoFar).replaceAllLiterally(replaceStr2, definedSoFar)
          if (replacement.contains(definedSoFar)) {
            // Add it to the end. it may have data dependencies in other fields
            orderedFieldTypes ::: List(ujson.read(replacement))
          } else {
            //Add it to the beginning, there shouldn't be any data dependencies on non-initialized types
            List(ujson.read(replacement)) ::: orderedFieldTypes
          }
        }
      })
      //upickle/ujson is mutable
      inputObject("fields") = replacementArr
      inputObject
    }
    val reorganizedSchema = recordsInF.toList.init.foldRight(
      ujson.read(
        recordsInF.last.toString
          .replaceAllLiterally(replaceStr, ujson.Js.JsonableSeq(names).toString)
          .replaceAllLiterally(replaceStr2, ujson.Js.JsonableSeq(names).toString)))({
      case (objectToUnion, accumulatedSchema) =>
        replacePlaceholderTermsInFieldsWithUnion(objectToUnion, accumulatedSchema)

    })
    new SchemaFor[Fix[F]] {
      override def schema: Schema =
        new Schema.Parser().parse(reorganizedSchema.toString().replaceAllLiterally(s"__$replaceName", ""))
    }
```

Is there a better way to do this? I haven't really messed with LabelledGeneric enough to know if this is implementable in shapeless instead of the way I did it, but I'm open to feedback about how to make this less horrible.

**Comment by DhashS** [comment]
It's also worth noting that this solution is not able to work when there are multiple recursive calls (like in the example I gave)

**Comment by sksamuel** [comment]
In your 2nd example there is no recursion. Did you mean to use Expr[T] ?

**Comment by DhashS** [comment]
No, since you'd substitute `Expr[T]` in the place of `T`, giving you something of the type `Expr[Expr[T]]`, with as many levels deep as your tree is. You'd then use the Fixpoint combinator (https://en.wikipedia.org/wiki/Fixed-point_combinator) to recover something of the type `Fix[Expr]`, implemented using primitive recursion.

Check out https://medium.com/disney-streaming/fix-point-type-for-gadt-scala-dc4e2cde349b for a more thorough rundown. 

Additionally, the code that actually works for me looks like

```scala
/*
  This implicit conversion enables Fixpoint type serialzation.
  Avro does not allow mutually recursive type definitions (https://issues.apache.org/jira/browse/AVRO-530)
  Avro does not allow named unions (https://issues.apache.org/jira/browse/AVRO-248).
  Using the approach in AVRO-530 as a temporary workaround, we reorder the schema to have no forward dependencies
  TODO this does not support multiple recursive calls (i.e. Mul[T](lhs: T (1), rhs: T (2)) extends Expr[T], only one
   */
  implicit def SchemaForFix[F[_]: Functor](implicit ev: SchemaFor[F[REPLACE.type]]): SchemaFor[Fix[F]] = {
    val replaceNamespace = REPLACE.getClass.getName.split('$').head
    val replaceName      = REPLACE.getClass.getSimpleName.replaceAllLiterally("$", "")
    val recordsInF       = read(SchemaFor[F[REPLACE.type]].schema.toString).arr
    val names            = recordsInF.toList.map(_("name").str).map(_.toString)
    val namespaces       = recordsInF.toList.map(_("namespace").str).map(_.toString)
    //One of the forms REPLACE can take, since it can get serialized as a case class by avro4s for some reason
    val replaceStr =
      s"""{"type":"record","name":"$replaceName","namespace":"$replaceNamespace","fields":[]}""".stripMargin
    //The straightforward type replace
    val replaceStr2 =
      s""""$replaceNamespace.$replaceName"""".stripMargin
    def replacePlaceholderTermsInFieldsWithUnion(inputObject: Js, accum: Js): Js = {
      // To figure out where to splice into the list
      val lastName = names.indexOf(accum("name").str)
      // package up the union type, with things to be declared, the struct defn, and then defns from inside the struct defn
      val endDefns = accum.toString + ",\"" + names
        .drop(lastName + 1)
        .zip(namespaces)
        .map(tup => tup._2 + "." + tup._1)
        .mkString("\",\"")
      val definedSoFar = endDefns match {
        case x if x.endsWith(",\"") =>
          "[\"" + names
            .take(lastName)
            .zip(namespaces)
            .map(tup => tup._2 + "." + tup._1)
            .mkString("\",\"") + "\"," + accum.toString + "]"
        case els =>
          "[\"" + names
            .take(lastName)
            .zip(namespaces)
            .map(tup => tup._2 + "." + tup._1)
            .mkString("\",\"") + "\"," + els + "\"]"
      }

      val replacementArr = inputObject("fields").arr.toList.foldLeft(List.empty[Js.Value])({
        case (orderedFieldTypes, field) => {
          val replacement =
            field.toString.replaceAllLiterally(replaceStr, definedSoFar).replaceAllLiterally(replaceStr2, definedSoFar)
          if (replacement.contains(definedSoFar)) {
            // Add it to the end. it may have data dependencies in other fields
            orderedFieldTypes ::: List(read(replacement))
          } else {
            //Add it to the beginning, there shouldn't be any data dependencies on non-initialized types
            List(read(replacement)) ::: orderedFieldTypes
          }
        }
      })
      replacementArr.filter(_.toString.startsWith("\"")).toSet
      //upickle/ujson is mutable
      inputObject("fields") = replacementArr
      inputObject
    }
    val reorganizedSchema = recordsInF.toList.init.foldRight(
      read(
        recordsInF.last.toString
          .replaceAllLiterally(replaceStr, Js.JsonableSeq(names).toString)
          .replaceAllLiterally(replaceStr2, Js.JsonableSeq(names).toString)))({
      case (objectToUnion, accumulatedSchema) =>
        replacePlaceholderTermsInFieldsWithUnion(objectToUnion, accumulatedSchema)

    })
    val topNodeName = reorganizedSchema("name").str
    val result =
      Js.JsonableSeq(
          List(reorganizedSchema) ::: names
            .zip(namespaces)
            .filterNot(_._1 == topNodeName)
            // fully qualify all names
            .map(tup => Js.Str(tup._2 + "." + tup._1)))
        .toString
        .replaceAllLiterally(s"__$replaceName", "")

    new SchemaFor[Fix[F]] {
      override def schema: Schema =
        new Schema.Parser().setValidate(true).parse(result)
    }
  }
```

**Comment by sksamuel** [comment]
I'm happy for you to make a PR to add this functionality, but in all honestly I don't understand it, so can't add it myself.
